### ISSUE TITLE
Lock issues when running unit tests with Jest

### ISSUE THREAD

**Issue by v-beltran** [nonara/ts-patch#142]
Whenever I run my unit tests with Jest (with `--coverage`) and `ts-patch/compiler` set (live compiler), I get the following errors:

_Loading module ts-patch/compiler failed with error: Could not acquire lock to write file. If problem persists, run ts-patch clear-cache and try again._

or

_Loading module ts-patch/compiler failed with error: EPERM: operation not permitted, open '{PATH}\.cache\ts-patch\locks\f38788ba0e2006162426c2f9961a166c.lock'_

Not sure how to solve it, sometimes ts-patch cache-clear works, sometimes not.

The workaround I have for now is removing the `ts-patch/compiler` from the Jest configuration and just run `ts-patch install` before any Jest command.

**Comment by nonara** [comment]
Hi Victor. Thanks for the report!

I'd be glad to take a look. Can you make a reproduction repository so I can see what's happening?

**Comment by SlyryD** [comment]
I see this issue intermittently in my work repository when building all our packages on Windows. We have ~160 packages and use `lerna` to execute `gulp build` in each package in order (with parallelization). The `build` task calls `node root\node_modules\ts-patch\bin\tspc.js --project tsconfig-esm.json --skipLibCheck --sourceRoot root\modules\package-name`.

I realized while writing this that the issue could be because we run 2 builds in parallel:
1. With TypeScript's tsc that produces commonjs for node.js
2. With ts-patch's tspc that produces ESM for browser with some transformations

**Comment by nonara** [comment]
Thanks for the added detail, @SlyryD. 

> I realized while writing this that the issue could be because we run 2 builds in parallel:

Interestingly, that's the problem that locks are supposed to solve.

- https://github.com/nonara/ts-patch/blob/master/projects/core/src/utils/file-utils.ts

## Analysis

> Loading module ts-patch/compiler failed with error: Could not acquire lock to write file. If problem persists, run ts-patch clear-cache and try again.

This error is most likely due to one of two reasons:

**1. Process terminated forcefully before lock is released**

Locks are wrapped in a method similar to "context handlers" in python. It uses `finally` to release the file, so even ctrl-c (SIGINT) should usually execute the finally clause. 

If it was forcefully exited somehow from jest at just the right moment, it could potentially leave an artifact behind, but it seems less likely.

I mention this only because @v-beltran mentioned that sometimes clearing cache works. If you can determine that you have lockfile artifacts that persist after the application runs, that would help me get an idea of what's going on.

**2. The wait time isn't long enough**

We have lockfile wait time set to 2 seconds. This seems fairly high, but maybe there's an issue if its multithreading processes with jest. It could make sense to allow this to be configured via an env var, and maybe bump the default up a little higher.

> Loading module ts-patch/compiler failed with error: EPERM: operation not permitted, open '{PATH}.cache\ts-patch\locks\f38788ba0e2006162426c2f9961a166c.lock'

This one is unusual. 

@v-beltran I assume you put `{PATH}` in yourself and omitted the real path, or is that the exact copy of the error?

The main thing I can think here is that perhaps we're getting a collision by another thread right between the `waitForLockRelease` and `fs.writeFileSync`.

This would normally seem a little unusual, but since jest spawns up all its threads at once, I could see this. I originally designed the locking to mitigate another issue which was much less likely for collision. This was before we had the live compiler, which amplifies the likelihood of race issues considerably.

## Solutions

I think what we can do for now is:

1. Increase default wait time to 4_000
2. Add an env var override for wait time
3. Mention the env var override in the error message
4. Mitigate lock file acquisition 👇 

Example:

```
let attempts = 0;
const maxAttempts = 3;

// Re-attempt 3x
while (attempts < maxAttempts) {
  try {
    waitForLockRelease(lockFilePath);
    fs.writeFileSync(lockFilePath, '');
    break; 
  } catch (error) {
    if (error.code === 'EPERM') {
      attempts++;
      if (attempts >= maxAttempts) {
        throw error; 
      }
    } else {
      throw error;
    }
  }
}
```

Short of any further detail that might help shed light on what's happening (or a repro), this is probably the best strategy for now

**Comment by SlyryD** [comment]
The issue doesn't appear with a persistent patch, so I'll try that out.

**Comment by witcher112** [comment]
Hi,

I've encountered a similar issue while using ts-patch in Rush monorepo (which by default uses parallelism for build and rebuild commands).

In my case it was:
```
Error: ENOENT: no such file or directory, unlink 'monorepo\common\temp\node_modules\.cache\ts-patch\locks\4d32208c952a47ac0ce2101df23df1ed.lock'     
    at Object.unlinkSync (node:fs:1884:11)
    at readFile (monorepo\common\temp\node_modules\.pnpm\ts-patch@3.2.1\node_modules\ts-patch\module\module-file.js:21:37)
    at getModuleFile (monorepo\common\temp\node_modules\.pnpm\ts-patch@3.2.1\node_modules\ts-patch\module\module-file.js:79:36)
    at getTsModule (monorepo\common\temp\node_modules\.pnpm\ts-patch@3.2.1\node_modules\ts-patch\module\ts-module.js:41:54)
  ...3 lines omitted...
    at Module._extensions..js (node:internal/modules/cjs/loader:1548:10)
    at Module.load (node:internal/modules/cjs/loader:1288:32)
```

I've fixed it by applying the following patch to `utils/file-utils.js` (wrapping unlinkSync in try...catch block):
```diff
diff --git a/utils/file-utils.js b/utils/file-utils.js
index 84e6028ae008ff2ee56cf2b4eb23764b28ef5ba9..801a209e504cb5630bd4cecd4d9a0a1ca0028806 100644
--- a/utils/file-utils.js
+++ b/utils/file-utils.js
@@ -70,7 +70,7 @@ function withFileLock(filePath, fn) {
     }
     finally {
         if (fs_1.default.existsSync(lockFilePath))
-            fs_1.default.unlinkSync(lockFilePath);
+            try { fs_1.default.unlinkSync(lockFilePath); } catch {}
     }
 }
 exports.withFileLock = withFileLock;

```

I would suggest using OS system locks instead as they guarantee to be atomic (in this case I believe two build processes got past `waitForLockRelease` at roughly the same time which resulted in the lockfile not being present when the second process finished thus making `unlinkSync` throw the error).

PS. Thanks for your amazing work on the package, I'm using it to integrate `typia` into my code :)

**Comment by witcher112** [comment]
Another thought - is there a reason to use file locking for `Live Compiler`?

Perhaps the easiest solution would be to disable it in this mode.

**Comment by nonara** [comment]
This may be fixed in the latest version (v3.3.0) due to swapping to rmSync (see: #164 #165)

If anyone wants to try and let me know, I'd appreciate it. If so, we'll close this out.

**Comment by SorsOps** [comment]
@nonara  I updated to latest, I haven't seen the bug manifest again on our github pipelines at all

**Comment by qlonik** [comment]
I'm still seeing this issue with `v3.3.0` of ts-patch. I'm running on the windows-based runner and I run jest with `jest --maxWorkers=50%`.

My error stack is 
```
FAIL src/config/BuildType.spec.ts
  ● Test suite failed to run

    Loading module ts-patch/compiler failed with error: Could not acquire lock to write file. If problem persists, run ts-patch clear-cache and try again.

            
      at Importer._import (node_modules/ts-jest/dist/utils/importer.js:159:19)
      at Importer.typescript (node_modules/ts-jest/dist/utils/importer.js:88:21)
      at new ConfigSet (node_modules/ts-jest/dist/legacy/config/config-set.js:209:51)
      at TsJestTransformer._createConfigSet (node_modules/ts-jest/dist/legacy/ts-jest-transformer.js:170:16)
      at TsJestTransformer._configsFor (node_modules/ts-jest/dist/legacy/ts-jest-transformer.js:148:34)
      at TsJestTransformer.getCacheKey (node_modules/ts-jest/dist/legacy/ts-jest-transformer.js:300:28)
      at ScriptTransformer._getCacheKey (node_modules/@jest/transform/build/ScriptTransformer.js:228:41)
      at ScriptTransformer._getFileCachePath (node_modules/@jest/transform/build/ScriptTransformer.js:289:27)
      at ScriptTransformer.transformSource (node_modules/@jest/transform/build/ScriptTransformer.js:525:32)
      at ScriptTransformer._transformAndBuildScript (node_modules/@jest/transform/build/ScriptTransformer.js:674:40)

```

EDIT: sorry for unnecessary spam. It seems I just needed to clear cache with `ts-patch clear-cache`.
